README.rst
==================
 Developer README
==================

Structure
---------

This tree contains two independent components:

- the SL compiler and library in ``slc/``. 
  For further information check the files ``slc/doc/README`` and
  ``slc/doc/HACKING``.

- the SL test suite in ``tests/``.

How to get started from this tree
---------------------------------

1. run ``bootstrap`` in ``slc/``

2. see the "Installation" steps in ``slc/doc/README``

Licensing
---------

This software is composite, with different parts licensed under
different terms. For more details see:

- ``tests/README``

- ``slc/doc/README.rst``, section "Licensing".



ChangeLog
2015-07-24  Raphael 'kena' Poss  <r.poss@uva.nl>

	Minor maintainer usability fix.

	* bootstrap: Recurse into `slc', so that `slc/tools/lib' and
	`slc/lib' get configured as well.

2012-11-02  Raphael 'kena' Poss  <r.poss@uva.nl>

	Minor packaging fix.

	* bootstrap: Avoid forcing Automake 1.11, since 1.12 is now out.

2012-04-30  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Add a top-level documentation.

	* README.rst: New.

2011-03-08  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Split programs, demos and benchmarks to a separate repository.

	* programs/: Remove. This is now hosted elsewhere.
	(https://git.svp-home.org/slprogs.git)

	* bootstrap,
	* configure.ac,
	* Makefile.am: Update accordingly.

2010-12-14  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Split deployment and installation tools to a separate repository.

	* deploy/: Remove. This is now hosted elsewhere.
	(https://git.svp-home.org/deploy.git)

	* bootstrap,
	* configure.ac,
	* Makefile.am: Update accordingly.

2010-11-29  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Update after CSA server migration.

	* deploy/installer/configure.ac: Adjust default URLs after the CSA
	server migration.

2010-08-24  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Update the installer after UvA move.

	* deploy/installer/configure.ac: Allow users to override the FETCH
	command. Use updated default site (dist.svp-home.org) for tags and
	distfiles.

2010-07-14  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Enhance the installer.

	* deploy/installer/slversion.in: New.
	* deploy/installer/configure.ac: Generate `slversion'.
	* deploy/installer/rules/slenv.mk: Install it.

	* deploy/installer/slenv.in: Sort the tags by mtime. Gracefully
	exit if no selection is made.

	* deploy/installer/README: Explain `deslenv' and `slversion'.

2010-07-11  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Minor installation fix.

	* deploy/installer/Makefile.am (install-data-local): Do not
	overwrite already installed tags, to preserve timestamps.

2010-06-30  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Fix `slenv'.

	* deploy/installer/slenv.in: Fix typo.

2010-06-24  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Minor installer update.

	* deploy/installer/configure.ac: Use Automake silent rules.

2010-06-24  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Minor installer update.

	* deploy/installer/slenv.in: Extend PYTHONPATH. 

2010-06-23  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Update installer version to suggest upgrading.

	* deploy/installer/build-aux/package-version: Bump to 3.1.

2010-06-23  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Extend the tool chain installer.

	* deploy/installer/configure.ac: Detect `bash', make fetch command
	silent and produce its output to stdout. Fetch and use tag files
	to determine version numbers automatically.
	* deploy/installer/Makefile.am: Use tags. New target `moretags' to
	fetch all tags available centrally. Clean up better on `make
	uninstall'.
	
	* deploy/installer/slenv,
	* deploy/installer/deslenv: Rename to...
	* deploy/installer/slenv.in,
	* deploy/installer/deslenv.in: ...here. Use @BASH@. Use tags.

	* deploy/installer/configure.ac: Generate `slenv' and `deslenv'.
	* deploy/installer/rules/slenv.mk: Do not distribute them.

	* deploy/installer/rules/mgsim.mk: Avoid creating `sl-current'
	since the shortcut is now provided by tag files.

2010-06-13  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Minor packaging fix.

	* deploy/installer/configure.ac: Require Autoconf 2.65, since we
	are using `m4_ifblank'.

2010-06-11  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Minor packaging fix.

	* bootstrap: Properly recurse into `deploy/installer'. Create
	`data' subdirectory for livermore benchmarks before generating
	data files.

2010-06-02  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Minor packaging fix.

	* deploy/dist/mgsim-dist.mk: Update check for `configure' failure.

2010-05-31  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Update simulator version in installer.

	* deploy/installer/versions.m4: Bump version number.

2010-05-31  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Update installer version to suggest upgrading.

	* deploy/installer/build-aux/package-version: Bump to 3.0.

2010-05-31  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Upgrade environment scripts.

	* deploy/installer/slenv: Also set `PYTHONPATH'.

	* deploy/installer/deslenv: New script to unset variables.
	* deploy/installer/rules/slenv.mk: Install the new script.

2010-05-31  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Change installer defaults to the new ABI.

	* deploy/installer/versions.m4: Bump version numbers.

2010-05-28  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Modify the stock GCC to better support C99's VLA's.

	The stock GNU compiler needs to juggle around the stack pointer
	when using large frames. When restoring the frame pointer on exit,
	it needs a temporary register. However at this point it does not
	honor `-ffixed', so a "forbidden" register could get used. This
	changeset prevents this by using the AT register instead.
	
	* deploy/patches/gcc.patch: Extend changes to `alpha.c'.

2010-04-29  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Update installer version to suggest upgrading.

	* deploy/installer/build-aux/package-version: Bump to 2.4.

2010-04-29  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Change installer defaults to the new ABI.

	* deploy/installer/versions.m4: Bump versions numbers.

2010-04-29  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Update copyright notice of files changed in 2010.	

	* bootstrap,
	* configure.ac,
	* Makefile.am,
	* deploy/installer/bootstrap: Add 2010 to copyright notice.

2010-04-26  Raphael 'kena' Poss  <r.c.poss@uva.nl>

	Enhance the deployment installer.

	* deploy/installer/Makefile.am,
	* deploy/installer/rules/mgsim.mk: Use `slreqs-$(SLNAME)' as base
	prefix for requirement, so that different toolchain versions can
	have different default prerequisites.

	* deploy/installer/slenv: Environment variables to force

lib/div/regs2.h
//                                                             -*- C++ -*-
// regs2.h: this file is part of the slc project.
//
// Copyright (C) 2009 The SL project.
// All rights reserved.
//
// $Id$
//
#define PHYS_T0 "$1"
#define PHYS_T1 "$2"
#define PHYS_T2 "$3"
#define PHYS_T3 "$4"
#define PHYS_T4 "$5"
#define PHYS_T5 "$6"
#define PHYS_T6 "$7"
#define PHYS_T7 "$8"
#define PHYS_T8 "$22"
#define PHYS_T9 "$23"
#define PHYS_T10 "$24"
#define PHYS_T11 "$25"

// Assembly scratch register
#define PHYS_AT "$28"

// Return value register
#define PHYS_V0 "$0"

// Incoming function parameters
#define PHYS_A5 "$21"
#define PHYS_A4 "$20"
#define PHYS_A3 "$19"
#define PHYS_A2 "$18"
#define PHYS_A1 "$17"
#define PHYS_A0 "$16"

// Pointer to current function
#define PHYS_PV "$27"
#define PHYS_T12 PHYS_PV

// Callee-save registers
#define PHYS_S0 "$9"
#define PHYS_S1 "$10"
#define PHYS_S2 "$11"
#define PHYS_S3 "$12"
#define PHYS_S4 "$13"
#define PHYS_S5 "$14"

// Return address
#define PHYS_RA "$26"

// Frame-pointer / callee-save
#define PHYS_FP "$15"
#define PHYS_S6 PHYS_FP

// Global pointer
#define PHYS_GP "$29"
// Stack pointer
#define PHYS_SP "$30"
// Zero
#define PHYS_ZERO "$31"


///////////////////////////////////
// Microgrids registers

// Formula
// N = shareds
// M = globals
// #L (locals) = (31 - 2 * N - M)
// #F (fixed locals during create) = 2 + N + M
// #R = #L - #F

// Example:
// N = 3, M = 8 -> #L = 17, #F = 13, #R = 4
// N = 3, M = 7 -> #L = 18, #F = 12, #R = 6
// N = 3, M = 6 -> #L = 19, #F = 11, #R = 8
// we choose the latter


#define RL0_N 0
#define RL1_N 1
#define RL2_N 2
#define RL3_N 3
#define RL4_N 4
#define RL5_N 5
#define RL6_N 6
#define RL7_N 7
#define RL8_N 8
#define RL9_N 9
#define RL10_N 10
#define RL11_N 11
#define RL12_N 12
#define RL13_N 13
#define RL14_N 14
#define RL15_N 15
#define RL16_N 16
#define RL17_N 17
#define RL18_N 18



#ifdef MT_SYMBOLIC
#define RG0 "$g0"
#define RG1 "$g1"
#define RG2 "$g2"
#define RG3 "$g3"
#define RG4 "$g4"
#define RG5 "$g5"

#define RS0 "$s0"
#define RS1 "$s1"
#define RS2 "$s2"

#define RD0 "$d0"
#define RD1 "$d1"
#define RD2 "$d2"

#define RL0 "$l0"
#define RL1 "$l1"
#define RL2 "$l2"
#define RL3 "$l3"
#define RL4 "$l4"
#define RL5 "$l5"
#define RL6 "$l6"
#define RL7 "$l7"
#define RL8 "$l8"
#define RL9 "$l9"
#define RL10 "$l10"
#define RL11 "$l11"
#define RL12 "$l12"
#define RL13 "$l13"
#define RL14 "$l14"
#define RL15 "$l15"
#define RL16 "$l16"
#define RL17 "$l17"
#define RL18 "$l18"

#define RZ "$zero"
#define RGP "$lgp"
#define RSP "$lsp"
#define RRA "$lra"
#define RPV "$lpv"
#define RRV "$lrv"
#define RFP "$lfp"

#define RSYNC "$sync"
#define RBRK "$brk"

#define RAS0 "$as0"
#define RAS1 "$as1"
#define RAS2 "$as2"

#define RAG0 "$ag0"
#define RAG1 "$ag1"
#define RAG2 "$ag2"
#define RAG3 "$ag3"
#define RAG4 "$ag4"
#define RAG5 "$ag5"

#else // !MT_SYMBOLIC

// Use argument registers for fixeds
#define RG0 PHYS_S5
#define RG1 PHYS_S4
#define RG2 PHYS_S3
#define RG3 PHYS_S2
#define RG4 PHYS_S1
#define RG5 PHYS_S0
#define RS0 PHYS_A0
#define RD0 PHYS_A1
#define RS1 PHYS_A2
#define RD1 PHYS_A3
#define RS2 PHYS_A4
#define RD2 PHYS_A5
// special
#define RZ PHYS_ZERO


// Match special with special
// at the end of the locals
#define RL18 PHYS_SP
#define RL17 PHYS_GP
#define RL16 PHYS_PV
#define RL15 PHYS_RA
#define RL14 PHYS_FP
#define RL13 PHYS_V0
#define RSP RL18
#define RGP RL17
#define RPV RL16
#define RRA RL15
#define RFP RL14
#define RRV RL13

// Allocate remaining locals
#define RL12 PHYS_AT
#define RL11 PHYS_T11
#define RL10 PHYS_T10
#define RL9 PHYS_T9
#define RL8 PHYS_T8
#define RL7 PHYS_T7
#define RL6 PHYS_T6
#define RL5 PHYS_T5
#define RL4 PHYS_T4
#define RL3 PHYS_T3
#define RL2 PHYS_T2
#define RL1 PHYS_T1
#define RL0 PHYS_T0

// Use locals for specials
#define RSYNC RL10
#define RBRK RL11

#define RAS0 RL7
#define RAS1 RL8
#define RAS2 RL9
#define RAG0 RL1
#define RAG1 RL2
#define RAG2 RL3
#define RAG3 RL4
#define RAG4 RL5
#define RAG5 RL6

#endif

#define RAS0_N RL7_N
#define RAS1_N RL8_N
#define RAS2_N RL9_N
#define RAG0_N RL1_N
#define RAG1_N RL2_N
#define RAG2_N RL3_N
#define RAG3_N RL4_N
#define RAG4_N RL5_N
#define RAG5_N RL6_N

// Fix registers; this effectively prevents
